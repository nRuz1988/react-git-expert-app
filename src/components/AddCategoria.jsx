
import { useState } from 'react';

/*Opcion 1
{setCategoria} => Se hace la desestructuraciÃ³n de la funcion setCategoria()
export const AddCategoria = ({setCategoria}) => {*/
export const AddCategoria = ({onNewCategoria}) => {

    const [inputValue,setInputValue] = useState('');

    const onInputChange = ({target}) =>{
        setInputValue(target.value);
    }
    const onSubmit = (event) =>{
        event.preventDefault();

        //Valida que en input digite mas de 1 caracter
        if(inputValue.trim().length <= 1) return;
        /*Opcion 1
        categorias => Es un callback  de UseState 
        setCategoria(categorias => [inputValue,...categorias]);*/

        onNewCategoria(inputValue.trim().toUpperCase());

        //setea la campo
        setInputValue('');
    }


    return (
        <form onSubmit={event => onSubmit(event)}>
            <input
                type="text"
                placeholder="Buscar"
                value = {inputValue}
                onChange={event => onInputChange(event)}
            />
        </form>
    )
}