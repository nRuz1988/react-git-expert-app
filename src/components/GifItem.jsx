
import React from 'react'
import propTypes from 'prop-types';

export const GifItem = ({id,titulo,url}) => {
  return (
    <div className='card'>
        <p>{titulo}</p>
        <img 
            src ={url}
            alt={titulo}
        />
    </div>
  )
}


/*Si desde el archivo padre(main) no envia el Props titulo  este sentecia se lo 
asigna por defauld*/
GifItem.defaultProps = {
  titulo : 'No existe titulo',
  url    : '',
}

//Sirve para validar el tipo y valor del campos (prototype)
GifItem.propTypes ={
  id     : propTypes.string,
  titulo  : propTypes.string.isRequired,
  url     : propTypes.string.isRequired,
}