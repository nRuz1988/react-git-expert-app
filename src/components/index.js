
export * from './AddCategoria';
export * from './GifItem';
export * from './GifGrid';