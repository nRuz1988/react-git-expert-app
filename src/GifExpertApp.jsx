import { useState } from 'react';
import { AddCategoria,GifGrid } from './components';



export const GifExpertApp = () => {

    const [categorias,setCategoria] = useState([]);

    const onAddCategoria = (newCateria)=>{

        //Si ya existe el valor en el arreglo, que no lo agregue
        if(categorias.includes(newCateria)) return;

        setCategoria([newCateria,...categorias]);
    }

    return (
        <>
            <h1>GifExpertApp</h1>
            <AddCategoria 
                    /* Opcion 1
                    Se envia la funcion setCategoria() como parametro
                    setCategoria= {setCategoria}*/
                    onNewCategoria={(value) => onAddCategoria(value)}
            />
                {
                    categorias.map(categoria =>(
                            <GifGrid 
                                key={categoria} 
                                categoria = {categoria}
                            />
                    ))
                }
        </>
    )
}