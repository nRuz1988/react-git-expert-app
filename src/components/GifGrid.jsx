
import {GifItem} from './GifItem';
import {useFechGifs} from '../hooks/useFechGifs';
import propTypes from 'prop-types';

export const GifGrid = ({categoria}) => {

      /*Hooks => No es mas que una funcion que regresa algo,ya sea personlizada 
      o de nativas de react */ 
     const {imagenes, isLoading} = useFechGifs(categoria);   

    //<> => fragmento, sirve para  cologar mas de un objeto html 
    return (
       <>
            <h3>{categoria}</h3> 
            {
                /*
                    && => Es un and logico, es decir si isLoading es true cumple la condicion
                    y se muestra el h2 de lo contrario no hace nada 
                */
                isLoading && (<h2>Cargando....</h2>)
            }

            <div className='card-grid'>
                {
                    /*{...img} --> Esto Sirve para esparcir todas las 
                    propiedades, cuando son muchas*/ 
                    imagenes.map(( img ) =>(
                        <GifItem 
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div>
       </> 
    )
}

GifGrid.propTypes ={
    categoria  : propTypes.string.isRequired,
}