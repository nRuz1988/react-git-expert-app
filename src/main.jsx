import React from 'react'
import ReactDOM from 'react-dom/client'
import {GifExpertApp} from './GifExpertApp';
import './style.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  /*
    https://reactjs.org/docs/strict-mode.html
    strict-mode => Ayuda principalmente a identificar componente y cosas varias (solo se
      aplica en modo desarrollo) 
  */
  <React.StrictMode>
    <GifExpertApp />
  </React.StrictMode>
);