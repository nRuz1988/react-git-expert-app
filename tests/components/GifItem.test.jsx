import { render,screen } from "@testing-library/react";
import {GifItem}  from '../../src/components'

describe('Test Component GifItem ', ()=>{

    const id     = '123';
    const titulo  = 'test';
    const url     = 'http://test.com/img.jpg';
        
    test('test  function GifItem Snapshot', () =>{

        const {container} = render(<GifItem 
            id={id}
            titulo={titulo}
            url={url}
        />);
        //Modificar el Snapshot -> npm test -- -u
        expect(container).toMatchSnapshot();
    });

    test('test  function GifItem Validando url y titulo de la img', () =>{

        render(<GifItem id={id}titulo={titulo}url={url}/>);
        /*Esto es como console log que muestra todo contenido del render
        screen.debug();*/

        /*Opcion 1
            expect(screen.getByRole('img').src).toBe(url);*/

        //opcion 2
        const {url,alt} = screen.getByRole('img');
        expect(url).toBe(url);
        expect(alt).toBe(titulo);

    });
    test('test  function GifItem Exita titulo en el componente', () =>{   
        render(<GifItem id={id}titulo={titulo}url={url}/>);
        expect(screen.getByText(titulo)).toBeTruthy();
    }); 
});