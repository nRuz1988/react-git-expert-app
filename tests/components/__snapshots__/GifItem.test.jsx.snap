// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Component GifItem  test  function GifItem Snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <p>
      test
    </p>
    <img
      alt="test"
      src="http://test.com/img.jpg"
    />
  </div>
</div>
`;
